packages/xod-client/src/hinting/validators.internal.js

```diff
@@ -0,0 +1,486 @@
+      mergeAllDeepWithConcat,
+      R.map(fn => fn(patch, project, prevErrors))
+    )(nodeValidators)
+);
+
+// :: [LinkValidateFn] -> Patch -> Project -> Map PatchPath DeducedPinTypes -> Map PatchPath PatchErrors -> Map LinkId LinkErrors
+const validateLinks = R.curry(
+  (validators, patch, project, allDeducedPinTypes, prevErrors) =>
+    R.compose(
+      R.map(R.objOf('errors')),
+      R.map(link =>
+        R.compose(
+          mergeAllDeepWithConcat,
+          R.map(fn => fn(link, patch, project, allDeducedPinTypes, prevErrors))
+        )(validators)
+      ),
+      R.indexBy(XP.getLinkId),
+      XP.listLinks
+    )(patch)
+);
+
+// :: [NodeValidateFn] -> [PinValidateFn] -> Patch -> Project -> Map PatchPath PatchErrors -> { nodes: Map NodeId NodeErrors }
+const getNodeErrors = R.curry(
+  (nodeValidators, pinValidators, patch, project, prevErrors) =>
+    R.compose(R.objOf('nodes'), validateNodes(nodeValidators, pinValidators))(
+      patch,
+      project,
+      prevErrors
+    )
+);
+// :: [LinkValidateFn] -> Patch -> Project -> Map PatchPath DeducedPinTypes -> Map PatchPath PatchErrors -> { links: Map LinkId LinkErrors }
+const getLinkErrors = R.curry(
+  (linkValidators, patch, project, allDeducedPinTypes, prevErrors) =>
+    R.compose(R.objOf('links'), validateLinks(linkValidators))(
+      patch,
+      project,
+      allDeducedPinTypes,
+      prevErrors
+    )
+);
+
+// :: PatchPath -> [Patch] -> [Patch]
+const filterPatchAndDependentPatchesByPatchPath = R.curry(
+  (patchPath, patches) =>
+    R.when(
+      () => patchPath,
+      R.filter(
+        R.either(
+          XP.hasNodeWithType(patchPath),
+          R.pipe(XP.getPatchPath, R.equals(patchPath))
+        )
+      )
+    )(patches)
+);
+
+const propChildrenEmpty = R.propSatisfies(R.pipe(R.values, R.all(R.isEmpty)));
+const propEmpty = R.propSatisfies(R.isEmpty);
+
+// :: { errors: Map ErrorType [Error] } -> Boolean
+const haveNoErrors = R.either(propEmpty('errors'), propChildrenEmpty('errors'));
+
+// :: { pins: Map PinKey (Map ErrorType [Error]) } -> Boolean
+const haveNoPinErrors = R.either(
+  propEmpty('pins'),
+  R.either(
+    propChildrenEmpty('pins'),
+    R.compose(R.all(propChildrenEmpty('errors')), R.values, R.prop('pins'))
+  )
+);
+
+// :: { nodes: Map NodeId { errors: Map ErrorType [Error], pins: Map PinKey (Map ErrorType [Error]) } } -> Boolean
+const haveNoNodeErrors = R.either(
+  propEmpty('nodes'),
+  R.compose(
+    R.all(R.both(haveNoErrors, haveNoPinErrors)),
+    R.values,
+    R.prop('nodes')
+  )
+);
+
+// :: { links: Map LinkId { errors: Map ErrorType [Error] } } -> Boolean
+const haveNoLinkErrors = R.either(
+  propEmpty('links'),
+  R.compose(R.all(haveNoErrors), R.values, R.prop('links'))
+);
+
+// :: PatchErrors -> Boolean
+const haveNoAnyErrors = R.allPass([
+  haveNoErrors,
+  haveNoNodeErrors,
+  haveNoLinkErrors,
+]);
+
+// =============================================================================
+//
+// API
```

Comment:
The API functions short doc comments are much desired :pray:  I can’t understand all that nuances of generally/not-generally, policies, `callFnIfExist` O_o